libcaf_core/caf/flow/op/retry.test.cpp

```diff
@@ -0,0 +1,197 @@
+// This file is part of CAF, the C++ Actor Framework. See the file LICENSE in
+// the main distribution directory for license terms and copyright or visit
+// https://github.com/actor-framework/actor-framework/blob/master/LICENSE.
+
+#include "caf/test/fixture/deterministic.hpp"
+#include "caf/test/fixture/flow.hpp"
+#include "caf/test/scenario.hpp"
+#include "caf/test/test.hpp"
+
+#include "caf/log/test.hpp"
+
+using namespace caf;
+
+namespace {
+
+struct fixture : test::fixture::deterministic, test::fixture::flow {
+  // Similar to retry::subscribe, but returns a retry_sub pointer instead of
+  // type-erasing it into a disposable.
+  template <class T = int, class Predicate>
+  auto raw_sub(caf::flow::observable<int> in, caf::flow::observer<int> out,
+               Predicate predicate) {
+    using sub_t = caf::flow::op::retry_sub<T, Predicate>;
+    auto ptr = make_counted<sub_t>(coordinator(), out, std::move(predicate),
+                                   in);
+    out.on_subscribe(caf::flow::subscription{ptr});
+    return ptr;
+  }
+};
+
+WITH_FIXTURE(fixture) {
+
+SCENARIO("the retry operator retries observable multiple times") {
+  int retry_count = 3;
+  auto retry_hdl = [&retry_count](const caf::error& what) {
+    if (what == sec::invalid_request && retry_count > 0) {
+      --retry_count;
+      return true;
+    } else
+      return false;
+  };
+  GIVEN("an observable that always fails") {
+    WHEN("calling retry") {
+      THEN("the observer receives error after retrying 3 times") {
```

Comment:
```suggestion
      THEN("the observer receives an error after retrying 3 times") {
```

More importantly, nothing in this test checks whether the retries actually happened or not.

Reply:
I had previously tested if the retries are happening by counting the elements in output. I have added more explicit tests to check for retries now.